### PR TITLE
API: register NEP 35 functions as array_functions

### LINKED ISSUE #27451: BUG: overridable functions from NEP 35 do not show up in `ARRAY_FUNCTIONS`
### Describe the issue:

Functions that were made dispatchable via NEP 35 (which added the `like` keyword argument) do not appear to be so using the `np.testing.overrides` API.

Taking for instance [np.asarray](https://numpy.org/doc/stable/reference/generated/numpy.asarray.html#numpy-asarray), whose `like` argument is documented as

> If an array-like passed in as like supports the `__array_function__` protocol, the result will be defined by it

Meanwhile, `np.testing.overrides.get_overridable_numpy_array_functions`'s docstring says
> List all numpy functions overridable via `__array_function__`

Since there's an API to check NEP 18 support, there should probably be one for NEP 35 too. I'm not sure wether the best route would be to extend the existing API to cover NEP 35 functions too, or if another set of functions should be added to `np.testing.overrides`.

### Reproduce the code example:

```python
import numpy as np
assert np.asarray in np.testing.overrides.get_overridable_numpy_array_functions()
```

### Error message:

```shell
Traceback (most recent call last):
  File "/Users/clm/dev/astropy-project/coordinated/astropy/t.py", line 2, in <module>
    assert np.asarray in np.testing.overrides.get_overridable_numpy_array_functions()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

### Python and NumPy Versions:

```
2.1.1
3.12.6 (main, Sep  9 2024, 21:36:32) [Clang 18.1.8 ]
```

### Runtime Environment:

```
[{'numpy_version': '2.1.1',
  'python': '3.12.6 (main, Sep  9 2024, 21:36:32) [Clang 18.1.8 ]',
  'uname': uname_result(system='Darwin', node='kwanzaabot.home', release='23.6.0', version='Darwin Kernel Version 23.6.0: Wed Jul 31 20:48:52 PDT 2024; root:xnu-10063.141.1.700.5~1/RELEASE_ARM64_T6020', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}}]
```

### Context for the issue:

We started discussing supporting a subset of these functions in `astropy.units` (https://github.com/astropy/astropy/issues/17001), but I haven't found an easy systematic way to discover the whole set (in particular because we're already testing for implementation completeness of NEP 18 functions).

### PR DESCRIPTION
**Pull request by neutrinoceros**
Close gh-27451
I aimed for a small footprint. If the additional responsibilities to `add_newdoc` are not acceptable it should be possible to avoid it, at the cost of more repetitions.
I also wasn't able to run the test suite locally yet, so I'm opening as a draft for now.

### CODE REVIEW

**Review by mhvk — COMMENTED** [review]
Looks quite nice! Since this all is (and was) a little fragile, I think it might be good to add a test that these functions are in fact in the list. Not completely sure what is the best place, but perhaps a new `np.testing.tests.test_overrides.py`?

**Comment by mhvk** [comment]
> I wasn't sure how to test this. Are you suggesting a test with a hard-coded list of expected results ?

No, more a spot check that, e.g., `np.asarray` is included. Basically to guard against the implementation being changed.

**Comment by neutrinoceros** [comment]
Gotcha. I did something in between. Let me know if you think it's too much.

**Review by mhvk — APPROVED** [review]
Nice! I'll approve but will let @seberg have a look too.

File: numpy/_core/function_base.py [hunk 1]
```diff
@@ -533,6 +533,8 @@ def add_newdoc(place, obj, doc, warn_on_python=True):
     """
     new = getattr(__import__(place, globals(), {}, [obj]), obj)
     if isinstance(doc, str):
+        if "${ARRAY_FUNCTION_LIKE}" in doc:
+            doc = overrides.get_array_function_like_doc(new, doc)
```
Comment (seberg) [inline numpy/_core/function_base.py:537]: Not sure I love the implicit nature.  I think I would prefer a kwarg for this.  But I suppose that may be a matter of opinion.

File: numpy/_core/overrides.py [hunk 2]
```diff
@@ -19,12 +19,13 @@
         compatible with that passed in via this argument."""
 )
 
-def set_array_function_like_doc(public_api):
-    if public_api.__doc__ is not None:
-        public_api.__doc__ = public_api.__doc__.replace(
-            "${ARRAY_FUNCTION_LIKE}",
-            array_function_like_doc,
-        )
+def get_array_function_like_doc(public_api, docstring_template=None):
```
Comment (seberg) [inline numpy/_core/overrides.py:22]: Would be nice to mention the registration in the name.  But if we don't have a quick idea, no worries.

**Review by seberg — COMMENTED** [review]
LGTM, but the test needs to move.  Other comments are tweaks, but don't have a quick idea on how and it's also fine as is.

File: numpy/_core/function_base.py (hunk 1 above)
Comment (neutrinoceros) [reply numpy/_core/function_base.py:537]: happy to go with your prefered style. What would you suggest we name the argument ?

File: numpy/testing/tests/test_overrides.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    }
+    assert nep35_python_functions.issubset(array_functions)
+
+    nep35_C_functions = {
+        np.arange, np.array, np.asanyarray, np.asarray,
+        np.ascontiguousarray, np.asfortranarray, np.empty,
+        np.frombuffer, np.fromfile, np.fromiter, np.fromstring,
+        np.zeros,
+    }
+    assert nep35_C_functions.issubset(array_functions)
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply numpy/testing/tests/test_overrides.py]: thanks ! indeed, the test can easily be rewritten (at least partially) in this module. I just moved it and rewrote it in a redundant style for now, as I see two ways to trim off the redundancy and would like your opinion first:
- only check for the subset that's already parametrized
- expand the existing parametrization to cover the whole (current) set
what option would you prefer ?

File: numpy/_core/overrides.py (hunk 2 above)
Comment (neutrinoceros) [reply numpy/_core/overrides.py:22]: agreed, I don't love that this function does two things, happy to take suggestions for a better name or pattern, but I don't have one now.

File: numpy/_core/function_base.py (hunk 1 above)
Comment (seberg) [reply numpy/_core/function_base.py:537]: Just go with something that seems sensible to you, I promise not to bike-shed ;).

Comment (neutrinoceros) [reply numpy/_core/function_base.py:537]: just to be clear though, did you mean a kwarg on `overrides.get_array_function_like_doc` of `add_newdoc` ? (I assume the former ?)

**Pull request merged**